src/coreclr/src/vm/method.cpp

```diff
@@ -4839,6 +4839,10 @@ bool MethodDesc::DetermineAndSetIsEligibleForTieredCompilation()
         // Functional requirement
         CodeVersionManager::IsMethodSupported(this) &&
 
+        // Functional requirement of being in COOP mode during a call count query.
+        // NativeCallable methods are, by definition, not initiated from COOP mode.
+        !HasNativeCallableAttribute() &&
```

Comment:
Is there a way to do this lazily? This going to make us lookup this NativeCallableAttribute for a lot.

Reply:
(Custom attribute lookup is not a cheap operation.)

Reply:
Agreed on cost. I ran it under a profiler and didn't observe any hits, but admit that doesn't mean it isn't expensive. Is this something that would better as a flag that we lookup and set once? When you say lazy do you mean closer to decision time? I thought this was that point.

Reply:
Well the issue I am seeing is one of truth. The `MethodDesc::IsEligibleForTieredCompilation()` call is made in many places, which means I could only update those locations that would respond to this attribute - I assume this is what is meant by lazy. However then other callers of this would be lied to since the flag in question was set. This leads me to believe that what we should do is look it up once and set the flag and never have to look it up again. Does that align with your concerns here?

Reply:
We are able to flags method as non-eligible for tiered compilation lazily when the method is JITed / executed for the first time (e.g. look for `JitSwitchedToOptimized()`). I think this should do something similar.

Reply:
Wow. There are lots of options here. I could also tell the JIT when it sees `JIT_FLAG_REVERSE_PINVOKE` to call back in and tell the runtime to reset the method to `CORINFO_FLG_SWITCHED_TO_OPTIMIZED`. This would move the query for `NativeCallableAttribute` to the absolute latest time. I will investigate a few avenues. Thanks!

Reply:
> tell the JIT when it sees JIT_FLAG_REVERSE_PINVOKE to call back in and tell the runtime to reset the method to CORINFO_FLG_SWITCHED_TO_OPTIMIZED

We try to avoid having runtime-specific policies like this hard-coded in the JIT.

Reply:
Okay. I will keep the policy side of things in the runtime. I have ideas that definitely avoid changes in the JIT.

Reply:
Okay. I will keep the policy side of things in the runtime. I have ideas that definitely avoid changes in the JIT.

Reply:
One way might be to pass `PrepareCodeConfig *` into `UnsafeJitFunction` instead of the `NativeCodeVersion` (which can be obtained from `PrepareCodeConfig`), and after the call to `GetCompileFlags()`, if `CORJIT_FLAG_REVERSE_PINVOKE` is set, remove the flag `CORJIT_FLAG_TIER0` and call `pConfig->SetJitSwitchedToOptimized()`, which would cause tiering to be disabled for the method in `JitCompileCodeLocked`.

Reply:
Yes, something like this should work. Also, we need to have an equivalent for R2R (this path won't kick in for R2R).

Reply:
Oh right. That could be done in `MethodDesc::GetPrecompiledCode`, something like this might work (the tier has to be changed before `SetNativeCode` to avoid races):
```c++
        pCode = GetPrecompiledR2RCode(pConfig);
        if (pCode != NULL)
        {
            LOG_USING_R2R_CODE(this);

        #ifdef FEATURE_TIERED_COMPILATION
            bool shouldTier = pConfig->GetMethodDesc()->IsEligibleForTieredCompilation();
        #if !defined(TARGET_X86) || !defined(TARGET_WINDOWS)
            if (shouldTier && HasNativeCallableAttribute())
            {
                // These methods are not tiered, update the tier in the code version
                NativeCodeVersion codeVersion = pConfig->GetCodeVersion();
                if (codeVersion.IsDefaultVersion())
                {
                    pConfig->GetMethodDesc()->GetLoaderAllocator()->GetCallCountingManager()->DisableCallCounting(codeVersion);
                }
                codeVersion.SetOptimizationTier(NativeCodeVersion::OptimizationTierOptimized);
                shouldTier = false;
            }
        #endif // !TARGET_X86 || !TARGET_WINDOWS
        #endif // FEATURE_TIERED_COMPILATION

            if (pConfig->SetNativeCode(pCode, &pCode))
            {
            #ifdef FEATURE_CODE_VERSIONING
                pConfig->SetGeneratedOrLoadedNewCode();
            #endif
            #ifdef FEATURE_TIERED_COMPILATION
                if (shouldTier)
                {
                    _ASSERTE(pConfig->GetCodeVersion().GetOptimizationTier() == NativeCodeVersion::OptimizationTier0);
                    pConfig->SetShouldCountCalls();
                }
            #endif
            }
        }
```

Reply:
It would be nice to avoid custom attribute lookups on the R2R path and encode this bit in the R2R fixup instead.

Reply:
@jkotas and @kouvel I have been able to apply the suggestion to `UnsafeJitFunction()`. I will need some guidance on the R2R side of things. I have found the `enum ReadyToRunFixupKind` in `readytorun.h`. Would the approach be to insert a bool into the readytorun metadata to indicate the `NativeCallableAttribute`?

An existing example or perhaps a pointer to some guide would be helpful here. My biggest issue is understanding how to query the R2R embedded data and exactly what data is interesting to embed.

Reply:
I'm not very familiar with how that should be implemented, @MichalStrehovsky would you be able to help?

Reply:
> I'm not very familiar with how that should be implemented, @MichalStrehovsky would you be able to help?

I haven't really done anything with R2R fixups and didn't spend much time reviewing that code either. Hopefully @trylek will be able to point you in the right direction. If not, I'll spend some time reading that code and come up with pointers.